### ISSUE TITLE
How does one download files?

### ISSUE THREAD

**Issue by dset0x** [poljar/weechat-matrix#161]
I must be really missing something here, so I apologize in advance.

Receiving a file results in a `exmc://` URI. How does one open that?

**Comment by emilazy** [comment]
Try the `contrib/matrix_decrypt` script; you'll probably need to pass a `--plumber` argument if you don't have [rifle](https://ranger.github.io/) installed.

**Comment by dset0x** [comment]
That works - thank you very much!

**Comment by MRAAGH** [comment]
Are there any recommended ways for integrating this `matrix_decrypt` script with Weechat-Matrix?

**Comment by poljar** [comment]
What do you mean by that? Just putting it inside your path somewhere is enough.

**Comment by emilazy** [comment]
It's quite annoying to have to manually copy the URL, open a terminal, and paste it after `matrix_decrypt`, especially when lots of images are sent in succession. The ideal thing would be to set up a URL handler for `exmc://` URLs on local clients, but unfortunately gnome-terminal fails to even display them as URLs for me, at least.

**Comment by poljar** [comment]
That does sound quite annoying. Termite recognizes them well, since many people use Weechat over tmux/screen, using the terminal to open the URLs seems to be the only way to handle this.

We can possibly modify the format those get printed out, to make gnome-terminal happy.

**Comment by dset0x** [comment]
@MRAAGH @emilazy What terminal are you using? The simplest workaround is to add `exmc://` to its recognized URIs and use ^L to switch to raw, then use whatever shortcut is available to you for opening the bottom-most (or so URI).

For rxvt-unicode specifically, I [patched the matcher extension](https://github.com/dset0x/rxvt-unicode/blob/matcher-delims/src/perl/matcher) so that it can reads URIs even when they are wrapped by weechat. . I haven't worked on it in a very long time, but it works for me (tm). Then have something like this in your `~/.Xresources`:
```
    URxvt.perl-ext-common: ...,matcher

    URxvt.matcher.pattern.0: (?:https?:\/\/|s?ftp:\/\/|news:\/\/|mailto:|file:\/\/|\bwww\.|irc:\/\/|thunderlink:\/\/|emxc:\/\/)[\\w\-\@;\\/?:&=%\$.+!*\\x27,~#]*(\\([\\w\-\\@;\\/?:&=%\\$.+!*\\x27,~#]*\\)|[\\w\\-\\@;\\/?:&=%\\$+*~])+

    URxvt.url-launcher  : xdg-open # in my case xdg-open is a wrapper for rifle
    URxvt.keysym.M-u:    perl:matcher:last
    URxvt.keysym.C-M-u:    perl:matcher:select

    URxvt.matcher.multiline: true
    URxvt.matcher.multiline.maxMatchLength: 500
    URxvt.matcher.multiline.delim.0.left:  (^|│)
    URxvt.matcher.multiline.delim.0.right: (\\\\|.($|│))
```

And then you end up with a nicely underlined URI:
![image](https://user-images.githubusercontent.com/8993055/75054329-28f1f680-54d3-11ea-8971-9c072f45108b.png)

@poljar I don't think you're doing anything wrong with the way you print it. Some people have some.. funny ideas about URIs in terminals, if you're interested in tracking: https://github.com/weechat/weechat/issues/1252

**Comment by poljar** [comment]
Hyperlink support in weechat would be awesome, especially if we're gonna need to print out even longer links... Though I think that has been blocked on ncurses adding support...

**Comment by MRAAGH** [comment]
Terminal shortcuts are nice, but in my opinion that misses out on "integration". When I have time, I might try to make a weechat plugin which grabs the last url in the current buffer and sends it to `matrix_decrypt`. Then it can really get a weechat keybinding!

**Comment by emilazy** [comment]
That would be great, especially if there's a way to open multiple images sent in quick succession.

**Comment by MRAAGH** [comment]
This probably comes down to personal preference, but the way I'd like it most is if all the mxid urls I receive were logged in a dedicated file, and I'd have a quick way of navigating that list and fetching any one of them, preferably from within Weechat itself. And the ability to save the files under their original filename (the name that came with the link).
I'm not promising anything from myself in the following weeks, though!

**Comment by bqv** [comment]
Could go further, have weechat automate this a bit an make a trigger to replace the exmc:// link with an actually accessible link

**Comment by jarukko** [comment]
I don't want to open a new issue about this same problem since I believe that this is user error. I can't figure out how to operate the decrypt script. 

I try to run it by calling it in other terminal `matrix_decrypt [emxc://matrix.org:443/_matrix/media/r0/download/matrix.org/example&hash=example&iv=example]` and I get the following error:
```
Traceback (most recent call last):
  File "/usr/bin/matrix_decrypt", line 81, in <module>
    main()
  File "/usr/bin/matrix_decrypt", line 55, in main
    if not query["key"] or not query["iv"] or not query["hash"]:
KeyError: 'iv'
```
What could be the cause for this?

**Comment by poljar** [comment]
I think your URL there is invalid, a query string is separated from the rest of the URL with a `?`, which seems to be missing here.
